Add some team related procedure blocks

Goldorion opened the pull request:
This PR adds some basic procedure blocks to manage teams. Only the basic procedure blocks (+ one team property) have been added to avoid a big PR.

I also added a new `Team management` nested category inside "World management" as team-related procedure blocks will contain quite a lot of procedure blocks, so we can keep categories with reasonable size.

Note: Currently, both getter and setter blocks are inside this category as it doesn't make sense yet to create another category. However, if this is needed in the next PR, they will be split.

Note 2: Teams support entities that are not players. However, other entities do not support things like friendly fire. A custom condition needs to be made if a user wants to disable this.

All added procedure blocks:
![image](https://github.com/MCreator/MCreator/assets/38427877/e5a7d422-209c-4a26-b125-8e02e1b12870)

KlemenDEV commented:
This looks a bit strange with the offset:

![image](https://github.com/MCreator/MCreator/assets/16374228/fa01824c-4524-486a-a0f7-4089f5d185ac)

It is also a bit confusing once one clicks to world management and the toolbox disappears:

![video](https://github.com/MCreator/MCreator/assets/16374228/2abca734-5619-4716-93ce-4d735670b6c3)

Goldorion commented:
This is how a nested category works. I can also disable the expanded by default of you think that'll be better.

KlemenDEV commented:
> This is how a nested category works

I am completely aware of this, I just said it looks a bit strange and could confuse users.

Spectrall368 commented:
> This is how a nested category works. It is handled by Blockly. I can also disable the expanded by default of you think that'll be better.

Goldorion could you add some procedures to change the team color too?

KlemenDEV commented:
> Goldorion could you add some procedures to change the team color too?

When you were asked to split PR into smaller pieces, it was for the same reason Goldorion decided to split the work into this PR and then into follow-up PRs. The (unwritten) rules of making PRs

Spectrall368 commented:
> > Goldorion could you add some procedures to change the team color too?
> 
> When you were asked to split PR into smaller pieces, it was for the same reason Goldorion decided to split the work into this PR and then into follow-up PRs. The (unwritten) rules of making PRs

I had imagined that if Goldorion had agreed to do those procedures he would have done it in another PR

Goldorion commented:
> The whole system is called "Scoreboard" (https://minecraft.fandom.com/wiki/Scoreboard), so it may make sense to call the category such and move other scoreboard blocks already existing there

I thought about this too, but I wasn't sure if it would have been a good idea as the other blocks have been designed to be made using entities, not scoreboard (which is also possible and surely produces a better code), so it can confuse some users.

Goldorion commented:
> > > Goldorion could you add some procedures to change the team color too?
> > 
> > 
> > When you were asked to split PR into smaller pieces, it was for the same reason Goldorion decided to split the work into this PR and then into follow-up PRs. The (unwritten) rules of making PRs
> 
> I had imagined that if Goldorion had agreed to do those procedures he would have done it in another PR

Planned procedure blocks are the following: `get/set player prefix, get/set display name, get/set color, get/set see friendly invisible, get/set name tag visibility, get/set death message visibility and get/set collision rule`. It's a total of 29 files and a lot more lines to review, so this PR + those extra blocks are way too many files and to review for a single PR (that can easily be small PRs)

Spectrall368 commented:
> > > > Goldorion could you add some procedures to change the team color too?
> > > 
> > > 
> > > When you were asked to split PR into smaller pieces, it was for the same reason Goldorion decided to split the work into this PR and then into follow-up PRs. The (unwritten) rules of making PRs
> > 
> > 
> > I had imagined that if Goldorion had agreed to do those procedures he would have done it in another PR
> 
> Planned procedure blocks are the following: `get/set player prefix, get/set display name, get/set color, get/set see friendly invisible, get/set name tag visibility, get/set death message visibility and get/set collision rule`. It's a total of 29 files and a lot more lines to review, so this PR + those extra blocks are way too many files and to review for a single PR (that can easily be small PRs)

Ok thanks for the answer

Goldorion commented:
>so it may make sense to call the category such and move other scoreboard blocks already existing there

The entity get/set scoreboard score blocks have been moved to the category (that was also renamed to Scoreboard management). I also changed the color of the set scoreboard score to match the other blocks. However, I didn't change the blocks' names, so we could avoid making a converter and potentially cause problems.

Goldorion commented:
> This looks a bit strange with the offset:
> 
> ![image](https://user-images.githubusercontent.com/16374228/272675753-fa01824c-4524-486a-a0f7-4089f5d185ac.png)

For the moment, I agree it can look weird, but if in the future we continue adding nested category instead ofadding main categories, it'll look less and less weird. Some current categories could also be moved to a nested category (e.g. `` Energy and fluid tanks` -> inside `Block management`, `Projectile procedures` -> `Entity management/data`, `Player data/management` -> `Entity data/management`)

KlemenDEV commented:
Fair point

KlemenDEV commented:
On hold until #4291

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/procedures/team_add_entity.java.ftl:
```diff
@@ -0,0 +1,7 @@
+{
```
Those additional braces outside are not needed as if already take care of wrapping inner variables, applies to most procedure blocks in the PR

KlemenDEV commented:
I would also make the category same color as other world procedures as well as blocks, same as how energy blocks have same color as other block procedures.

Let's leave more colors to the plugins

Goldorion commented:
Done

KlemenDEV commented on plugins/mcreator-core/procedures/team_add_entity.json:
```diff
@@ -0,0 +1,40 @@
+    ],
+    "dependencies": [
+      {
+        "name": "world",
```
World can be obtained from entity.level(), just save entity into tmp variable first so procedure providing entity is not called multiple times. Applies to other cases like this too.

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1125,6 +1126,14 @@ blockly.block.spawn_particle=Spawn single particle at x: %1 y: %2 z: %3 with vx:
+blockly.block.team_get_friendlyfire=Is friendly fire for team with name %1 enabled
+blockly.block.team_remove_entity=Remove entity %2 from team %1
+blockly.block.team_set_friendlyfire=Set friendly fire for team with name %1 to %2
+blockly.block.team_set_friendlyfire.tooltip=Friendly fire only works with players.
```
Since it is only one sentence, you can omit .

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/procedures/team_add_entity.java.ftl:
```diff
@@ -0,0 +1,6 @@
+{
+	Level _level = entity.level();
```
I meant to not use dependencies at all, since you already get ${input$entity}.

We no longer require entity as dependency but as input in all procedure blocks.

Goldorion replied:
Oh ok, sorry. I didn't understand correctly

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/procedures/team_add_entity.java.ftl:
```diff
@@ -0,0 +1,6 @@
+{
+	Level _level = ${input$entity}.level();
```
>  just save entity into tmp variable first so procedure providing entity is not called multiple times. Applies to other cases like this too.

Right now, entity is obtained twice, but could be just once into tmp variable. Matters eg. for in range procedure which is resource heavy

Also check 1.19.4, it has no .level() yet, only .level

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/procedures/team_add_entity.java.ftl:
```diff
@@ -0,0 +1,6 @@
+{
+	Entity _entity${cbi} = ${input$entity};
```
${cbi} is only needed for ouput / inline statements. Here you can use eg _entityTeamAdd

KlemenDEV commented on plugins/generator-1.20.1/forge-1.20.1/procedures/team_get_friendlyfire.java.ftl:
```diff
@@ -0,0 +1,9 @@
+(new Object() {
```
Those could probably be reduced to inline methods with ternary operator, applies to friendly fire and get entity team.

This way we avoid Object creation which is our last resort only when really needed

KlemenDEV replied:
Eg. entity_has_item_inventory

Goldorion replied:
This makes the code ugly, messy and just bad.
```java
((world instanceof Level _level) && _level.getScoreboard().getPlayerTeam(${input$name}) != null && _level.getScoreboard().getPlayerTeam(${input$name}).isAllowFriendlyFire())
```
or if you want to use `?` and `:`, it's even worse
 ```java
((world instanceof Level _level) ? _level.getScoreboard().getPlayerTeam(${input$name}) != null ? _level.getScoreboard().getPlayerTeam(${input$name}).isAllowFriendlyFire() : false : false)
```

Defeatomizer replied:
We already use that approach with some blocks AFAIK. About ternary operator, it is mostly needed when working with data other than boolean so not really needed in your case (on a side note, you could use `a & b ? c : false` rather than changing `a & b & c` directly to `a ? b ? c : false : false`).

KlemenDEV replied:
The performance burden of "new Object" allocating on heap and fragmentation heap will not be accepted.

You will need to change the way we use in other locations. Check existing code for inspiration

Goldorion commented on plugins/generator-1.19.4/forge-1.19.4/procedures/team_get_entity_team.java.ftl:
```diff
@@ -1 +1,2 @@
-(${input$entity}.level.getScoreboard().getPlayersTeam(${input$entity}.getStringUUID()) != null ? ${input$entity}.level.getScoreboard().getPlayersTeam(${input$entity}.getStringUUID()).getName() : "")
\ No newline at end of file
+(${input$entity} instanceof LivingEntity _teamEnt${cbi} && _teamEnt${cbi}.level.getScoreboard().getPlayersTeam(_teamEnt${cbi}.getStringUUID()) != null ?
```
By adding a check for living entities, you remove the possibility of adding non living entities to a team while it is supported by the game (e.g. armor stands)

KlemenDEV replied:
I see

KlemenDEV replied:
Seems ArmorStands are living entities too:

![image](https://github.com/MCreator/MCreator/assets/16374228/4d206dac-7913-49d8-9c40-f30d9bfc46a0)

Defeatomizer commented on plugins/generator-1.19.4/forge-1.19.4/procedures/team_get_entity_team.java.ftl:
```diff
@@ -1,2 +1,2 @@
-(${input$entity} instanceof LivingEntity _teamEnt${cbi} && _teamEnt${cbi}.level.getScoreboard().getPlayersTeam(_teamEnt${cbi}.getStringUUID()) != null ?
-	_teamEnt${cbi}.level.getScoreboard().getPlayersTeam(_teamEnt${cbi}.getStringUUID()).getName() : "")
\ No newline at end of file
+(${input$entity} instanceof LivingEntity _teamEnt && _teamEnt.level.getScoreboard().getPlayersTeam(_teamEnt.getStringUUID()) != null ?
```
What blocks do need this index then as I've seen other output blocks use it for pattern variables too?

KlemenDEV replied:
Blocks that can appear in || or &&
